File: node/kmsEnvelopeDecrypt.js
const crypto = require('crypto');
const { KMSClient, DecryptCommand } = require('@aws-sdk/client-kms');
const meow = require('meow');
const getStdin = require('get-stdin');

const b64Decode = (buffer) => Buffer.from(buffer, 'base64');

(async () => {
  const noStdin = process.stdin.isTTY;

  const cli = meow({
    flags: {
      key: {
        alias: 'k',
        isRequired: true,
      },
      secret: {
        alias: 's',
        isRequired: noStdin,
        default: noStdin ? undefined : await getStdin(),
      },
    },
  });

  const { key, secret } = cli.flags;

  const client = new KMSClient();

  const kmsKey = await client
    .send(
      new DecryptCommand({
        CiphertextBlob: b64Decode(key),
      })
    )
    .then((r) => Buffer.from(r.Plaintext));

  const finalKey = b64Decode(secret);
  const iv = finalKey.slice(0, 16);
  const encryptedValue = finalKey.slice(16);

  const cipher = crypto.createDecipheriv('AES-256-CBC', kmsKey, iv);
  cipher.update(encryptedValue);
  const decrypted = cipher.final().toString();

  console.log(decrypted);
})();

File: python/kms-envelope-decrypt.py
import argparse
import sys
from base64 import b64decode

import boto3
from Crypto.Cipher import AES

no_stdin = sys.stdin.isatty()

parser = argparse.ArgumentParser()
parser.add_argument('--key', '-k', help="KMS Data Key to use to decrypt (base64)", required=True)
parser.add_argument('--secret', '-s', help="Secret to decrypt (base64). Will attempt to read from stdin", default=(None if no_stdin else sys.stdin), required=no_stdin)
args = parser.parse_args()

client = boto3.client('kms')

response = client.decrypt(
    CiphertextBlob=b64decode(args.key)
)
kms_key = response["Plaintext"]

final_key = b64decode(args.secret)

iv = final_key[:16]
encrypted_value = final_key[16:]

cipher = AES.new(kms_key, AES.MODE_CBC, iv=iv)
decrypted = cipher.decrypt(encrypted_value)

output = decrypted.decode("utf-8")
print(output)
